From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Praveen James <pjames@nvidia.com>
Date: Fri, 31 Mar 2023 03:38:11 +0000
Subject: [PATCH 059/240] NVIDIA: SAUCE: tegra-epl: send state notification fsi

tegra-epl driver updated to send
suspend and shutdown notification to fsi

bug 4053113

Change-Id: I550c1c899317de6b7aa29c41d46a881de1b57282
Signed-off-by: Praveen James <pjames@nvidia.com>
Reviewed-by: Sidharth Varier <svarier@nvidia.com>
Reviewed-by: Laxman Dewangan <ldewangan@nvidia.com>
Reviewed-on: https://git-master.nvidia.com/r/c/linux-stable/+/2902846
Tested-by: Abhilash G <abhilashg@nvidia.com>
Reviewed-by: Abhilash G <abhilashg@nvidia.com>

diff --git a/drivers/platform/tegra/tegra-epl.c b/drivers/platform/tegra/tegra-epl.c
index e45195cbc3600..d7b7cae727906 100644
--- a/drivers/platform/tegra/tegra-epl.c
+++ b/drivers/platform/tegra/tegra-epl.c
@@ -1,5 +1,5 @@
 // SPDX-License-Identifier: GPL-2.0
-// Copyright (c) 2021-2022, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+// Copyright (c) 2021-2023, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
 
 #include <linux/module.h>
 #include <linux/fs.h>
@@ -27,6 +27,13 @@
 /* Error index offset in mission status register */
 #define ERROR_INDEX_OFFSET	24U
 
+/* signature code for HSP pm notify data */
+#define PM_STATE_UNI_CODE	0xFDEF
+
+/* State Management */
+#define PM_SUSPEND	6U
+#define PM_SHUTDOWN	PM_SUSPEND
+
 enum handshake_state {
 	HANDSHAKE_PENDING,
 	HANDSHAKE_FAILED,
@@ -202,6 +209,26 @@ int epl_report_error(struct epl_error_report_frame error_report)
 }
 EXPORT_SYMBOL_GPL(epl_report_error);
 
+static int epl_client_fsi_pm_notify(u32 state)
+{
+	int ret;
+	u32 pdata[4];
+
+	pdata[0] = PM_STATE_UNI_CODE;
+	pdata[1] = state;
+	pdata[2] = state;
+	pdata[3] = PM_STATE_UNI_CODE;
+
+	mutex_lock(&hs_state_mutex);
+	if (hs_state == HANDSHAKE_DONE)
+		ret = mbox_send_message(epl_hsp_v->tx.chan, (void *) pdata);
+	else
+		ret = -ENODEV;
+	mutex_unlock(&hs_state_mutex);
+
+	return ret < 0 ? ret : 0;
+}
+
 static int epl_client_fsi_handshake(void *arg)
 {
 	mutex_lock(&hs_state_mutex);
@@ -238,13 +265,15 @@ static int epl_client_fsi_handshake(void *arg)
 
 static int __maybe_unused epl_client_suspend(struct device *dev)
 {
+	int ret;
 	pr_debug("tegra-epl: suspend called\n");
 
+	ret = epl_client_fsi_pm_notify(PM_SUSPEND);
 	mutex_lock(&hs_state_mutex);
 	hs_state = HANDSHAKE_PENDING;
 	mutex_unlock(&hs_state_mutex);
 
-	return 0;
+	return ret;
 }
 
 static int __maybe_unused epl_client_resume(struct device *dev)
@@ -369,6 +398,20 @@ static int epl_client_probe(struct platform_device *pdev)
 	return ret;
 }
 
+static void epl_client_shutdown(struct platform_device *pdev)
+{
+	pr_debug("tegra-epl: shutdown called\n");
+
+	if (epl_client_fsi_pm_notify(PM_SHUTDOWN) < 0)
+		pr_err("Unable to send notification to fsi\n");
+
+	mutex_lock(&hs_state_mutex);
+	hs_state = HANDSHAKE_PENDING;
+	mutex_unlock(&hs_state_mutex);
+
+	epl_unregister_device();
+}
+
 static void epl_client_remove(struct platform_device *pdev)
 {
 	epl_unregister_device();
@@ -381,8 +424,9 @@ static struct platform_driver epl_client = {
 		.of_match_table = of_match_ptr(epl_client_dt_match),
 		.pm = pm_ptr(&epl_client_pm),
 	},
-	.probe			= epl_client_probe,
-	.remove			= epl_client_remove,
+	.probe		= epl_client_probe,
+	.shutdown	= epl_client_shutdown,
+	.remove		= epl_client_remove,
 };
 
 module_platform_driver(epl_client);
-- 
2.43.0

